crates/ark/src/analysis/parse_boundaries.rs

```diff
@@ -0,0 +1,375 @@
+//
+// parse_boundaries.rs
+//
+// Copyright (C) 2024 Posit Software, PBC. All rights reserved.
+//
+//
+
+use harp::vector::CharacterVector;
+use harp::vector::Vector;
+use harp::ParseResult;
+use harp::RObject;
+
+/// Boundaries are ranges over lines of text.
+#[derive(Debug)]
+pub struct ParseBoundaries {
+    pub complete: Vec<std::ops::Range<usize>>,
+    pub incomplete: Option<std::ops::Range<usize>>,
+    pub error: Option<std::ops::Range<usize>>,
```

Comment:
I'm just now realizing a better name for this is probably really `LineRange` (`TextRange` is for a byte range in a document)

The fact that it took me until the test section of the PR to figure this out probably does mean a named wrapper like this would be very helpful

```rust
struct LineRange {
  start: usize,
  end: usize,
}
```

Reply:
You're right documenting the contents with a type name would go a long way here.

`ArkRange` is Ark specific rather thant LSP- or TS-specific. It uses the row/col approach instead of byte offsets.

For byte offsets and ranges we may want to use https://github.com/rust-analyzer/text-size/

But that's storing u32 instead of usize, which makes it harder to interface with the rest of Rust, and we might not need

Reply:
These types make it harder to provide generic algorithms, e.g. for `fn merge_overlapping<T>(ranges: Vec<std::ops::Range<T>>) -> Vec<std::ops::Range<T>>`. Not that we really need the genericity now...

Reply:
ah but the ruff types are directly imported from rust-analyzer's text-size crate!